app/src/main/java/com/babylon/wallet/android/data/repository/state/StateRepository.kt

```diff
@@ -0,0 +1,236 @@
+
+            resource.copy(items = allNewItems)
+        }
+    }
+
+    override fun observeResourceDetails(resourceAddress: String, accountAddress: String?): Flow<Resource> = flow {
+        val cachedEntity = stateDao.getResourceDetails(
+            resourceAddress = resourceAddress,
+            minValidity = resourcesCacheValidity()
+        )
+        val amount = accountAddress?.let {
+            stateDao.getAccountResourceJoin(resourceAddress = resourceAddress, accountAddress = accountAddress)?.amount
+        }
+
+        val cachedResource = cachedEntity?.toResource(amount)
+        if (cachedResource != null) {
+            emit(cachedResource)
+        }
+
+        if (cachedResource?.isDetailsAvailable == false) {
+            val item = stateApi.getSingleEntityDetails(
+                address = resourceAddress,
+                metadataKeys = setOf(
+                    ExplicitMetadataKey.NAME,
+                    ExplicitMetadataKey.SYMBOL,
+                    ExplicitMetadataKey.DESCRIPTION,
+                    ExplicitMetadataKey.RELATED_WEBSITES,
+                    ExplicitMetadataKey.ICON_URL,
+                    ExplicitMetadataKey.INFO_URL,
+                    ExplicitMetadataKey.VALIDATOR,
+                    ExplicitMetadataKey.POOL,
+                    ExplicitMetadataKey.TAGS,
+                    ExplicitMetadataKey.DAPP_DEFINITIONS
+                )
+            )
+            val updatedEntity = stateDao.updateResourceDetails(item)
+
+            emit(updatedEntity.toResource(amount))
+        }
+    }
+
+    override suspend fun getNFTDetails(resourceAddress: String, localId: String): Result<Resource.NonFungibleResource.Item> =
+        withContext(dispatcher) {
+            val cachedItem = stateDao.getNFTDetails(resourceAddress, localId, resourcesCacheValidity())
+
+            if (cachedItem != null) {
+                return@withContext Result.success(cachedItem.toItem())
+            }
+
+            stateApi.nonFungibleData(
+                StateNonFungibleDataRequest(
+                    resourceAddress = resourceAddress,
+                    nonFungibleIds = listOf(localId)
+                )
+            ).toResult().mapCatching { response ->
+                val item = response.nonFungibleIds.first()
+                val entity = item.asEntity(resourceAddress, InstantGenerator())
+                stateDao.insertNFTs(listOf(entity))
+                entity.toItem()
+            }
+        }
+
+    override suspend fun getOwnedXRD(accounts: List<Network.Account>): Result<Map<Network.Account, BigDecimal>> =
+        accountsStateCache.getOwnedXRD(accounts = accounts)
+
+    override suspend fun getEntityOwnerKeys(entities: List<Entity>): Result<Map<Entity, OwnerKeyHashesMetadataItem>> = runCatching {
+        if (entities.isEmpty()) return@runCatching mapOf()
+
+        val entitiesWithOwnerKeys = mutableMapOf<Entity, OwnerKeyHashesMetadataItem>()
```

Comment:
Entity is not a data class, so I don't think you can safely use it as a key of has map. I would rather use `entity.address` here

Reply:
Isn't Entity though overridden from Account and Persona? Won't their equals and hash functions suffice at runtime?

Reply:
tested and indeed it is here, I could swear that I had such issue before, not sure maybe I was not using `sealed class` then